kombu/common.py

```diff
@@ -105,6 +105,12 @@ def declaration_cached(entity, channel):
 
 def maybe_declare(entity, channel=None, retry=False, **retry_policy):
     """Declare entity (cached)."""
+    if retry:
+        return _imaybe_declare(entity, channel, **retry_policy)
+    return _maybe_declare(entity, channel)
+
+
+def _maybe_declare(entity, channel):
     is_bound = entity.is_bound
```

Comment:
@georgepsarakis thanks for the response.
Assume the following chain of events:
1. `maybe_declare` [is called](https://github.com/celery/celery/blob/master/celery/backends/rpc.py#L168) like that: `maybe_declare(entity, retry=True)`. Notice that `channel=None`, which is the default.
2. `maybe_declare`'s [logic](https://github.com/celery/kombu/blob/master/kombu/common.py#L117) is to call `_imaybe_declare` with `channel=entity.channel`
3. `_imaybe_declare` calls `ensure(entity, _maybe_declare)` which [wraps](https://github.com/celery/kombu/blob/master/kombu/connection.py#L484) `_maybe_declare` in a function that ensures that `entity.channel` is working, and if not, fixing it.
4. `_imaybe_declare` then calls the wrapped func, [specifying the `channel` param directly](https://github.com/celery/kombu/blob/master/kombu/common.py#L147).
5. Now let's assume the connection is lost somewhere in the middle, and `entity.channel` is broken or None. [`_ensured`](https://github.com/celery/kombu/blob/master/kombu/connection.py#L484) will catch the exception, then [call `revive`](https://github.com/celery/kombu/blob/master/kombu/connection.py#L518) on entity, which tries to resolve that by opening a new connection, assigning it to `entity.channel`, and calling the wrapped function (`_maybe_declare`) again.
6. However, `_maybe_declare` **will not use the updated `entity.channel`**, as the `channel` parameter was passed directly by `maybe_declare` in step 4. In other words, `_maybe_declare` will get the broken channel as an argument over and over.

Does this make sense now?